### ISSUE TITLE
Wisp supports joint rendering of various pipelines (NeRF, SDFs, meshes)

### ISSUE THREAD

**Issue by 3a1b2c3** [NVIDIAGameWorks/kaolin-wisp#35]
Very curious if there are any examples of combining pipelines? Just add them to multiple scene_graph.neural_pipelines

**Comment by orperel** [comment]
Hi @3a1b2c3 , correct, adding objects to WispState's `scene_graph.neural_pipelines` should do the trick.
We'll add an example to showcase that soon.

Be aware that for now, the neural fields we optimize are assumed to exist in the normalized range of -1 to 1 (supporting per instance model matrix is on our short-term roadmap)

**Comment by 3a1b2c3** [comment]
Great, what template would be addong a mesh?

**Comment by orperel** [comment]
We don't have renderable meshes yet ;)

1. This is not mesh rendering per se, but a quick workaround - 
We have a PR under review for rendering kaolin's [Structured Point Clouds](https://kaolin.readthedocs.io/en/v0.9.1/modules/kaolin.ops.spc.html) as a neural field, which should be integrated very soon. That would entail converting your mesh to a SPC [like so](https://kaolin.readthedocs.io/en/v0.9.1/modules/kaolin.ops.spc.html?highlight=convert#conversions).

2. The correct way to go about it would be to implement a [RasterizedRenderer](https://github.com/NVIDIAGameWorks/kaolin-wisp/blob/df9e4d69b428f0262a82e28c2dc5830ba58a1e17/wisp/renderer/core/api/base_renderer.py#L148) for rendering meshes -> RenderBuffer.
For example, kaolin's [DIB-R rasterizer ](https://kaolin.readthedocs.io/en/latest/modules/kaolin.render.mesh.html#kaolin.render.mesh.dibr_rasterization) could be used as an actual implementation for such RasterizedRenderer.

**Comment by 3a1b2c3** [comment]
![image](https://user-images.githubusercontent.com/74843139/188128169-de7a0c5e-1c78-4272-a722-4359d0a2a925.png)
not nice code but etting closer

**Comment by 3a1b2c3** [comment]
Ugly code but got it in state
![image](https://user-images.githubusercontent.com/74843139/188128516-bc623dc8-ddb1-49a5-a20d-a4ed280b9070.png)

**Comment by 3a1b2c3** [comment]
Would love to see a road map for this project at some point?

**Comment by 3a1b2c3** [comment]
![image](https://user-images.githubusercontent.com/74843139/188310709-b89ab6c1-d870-44d6-bed1-bfdd7048f61b.png)
I get both data sets but only sdf one trains

**Comment by 3a1b2c3** [comment]
What is your preferred flavor of transform matrices, OpenGL style?

**Comment by 3a1b2c3** [comment]
Am i reading the code correctly that sdf can support texturing?

**Comment by 3a1b2c3** [comment]
Small win: getting to train nerf and mesh. Still fighting transforms a a bit, would you normalize sdf in first place if thats not th e final result?
![image](https://user-images.githubusercontent.com/74843139/189480797-e5f44c4e-68ce-403b-807f-cbacb2d9285c.png)

**Comment by orperel** [comment]
Hi @3a1b2c3 ,
Great to see you've got some progress there!

For your questions - 
1. "OptimizerApp wont allow multiple inputs for training function" - did you end up creating 2 separate trainers for nerf / sdf? If so you could try:
```
def joint_train_step():
   nerf_trainer.iterate()
   sdf_trainer.iterate()
```
and pass `trainer_step_func=joint_train_step` to the OptimizerApp.

2. "Can I use RasterizedRenderer and Nerf renderer at the same time?" - yup, that's the idea behind the renderers. They each represent an object in your scene graph so you can have multiple - a mesh and some nerf at the same time (even if one is ray traced and the other is rasterized). I've just added some explanations and diagrams about the renderers, you can check them out here: https://github.com/NVIDIAGameWorks/kaolin-wisp/tree/main/wisp/renderer

3. For roadmap of missing capabilities: right now we're still setting the course - but if you're worried about implementing something from our future plans, please don't be :) We're always happy to accept external contributions! (i.e. if you get support for meshes working).

4. Re: normalization - yes, the scale you set during optimization is assumed to be -1 to 1 (that's just convenient). What you're missing is an object transform that takes the rendered object and translates / rotates / scales it. I actually implemented, but I still need to iron out some bugs before I can push that one. See attachment:

https://user-images.githubusercontent.com/9556101/190153281-374dfd75-d0d9-4486-8928-968166c1bf7c.mp4

**Comment by 3a1b2c3** [comment]
Having the transform exposed in the gui is way nicer than in the config like i did it :). I grabbed some code from pytorch3d for that. 
I have to check my code my code for the training details, basically i allow an array of trainers and hook callbacks. There is some problem with stopiteration exception now i think. Thanks for your help

**Comment by 3a1b2c3** [comment]
Just trying the new version after going to Europe ... lovely 
![image](https://user-images.githubusercontent.com/74843139/194707652-afc6e135-f5cc-42cf-9ba3-ab4953d320b3.png)

**Comment by 3a1b2c3** [comment]
Would it make sense to add it to the config system?

**Comment by 3a1b2c3** [comment]
So I kind of could do afrankenrender  (2 renders needed) vs frankennerf (Merge trained nerf and mesh use 1 renderer) approach. Might be worth trying both?

**Comment by 3a1b2c3** [comment]
Nerfstudios renderer seems to be dealing with that scene slightly better (they also have controllable clipping planes)
![image](https://user-images.githubusercontent.com/74843139/198525519-b3f73931-c151-4d0d-9e7f-a92bc77b3aea.png)

**Comment by orperel** [comment]
Hi @3a1b2c3 !
Just to let you know that ObjectTransforms are now finally pushed:
https://github.com/NVIDIAGameWorks/kaolin-wisp/pull/65/files

How to use them:
You should see them right there in the gui.
You could also define a transform programmatically like so:

```
from wisp.renderer.core.api import add_to_scene_graph, remove_from_scene_graph
from wisp.core import ObjectTransform

t = ObjectTransform()
add_to_scene_graph(wisp_state, name=you_object_name, obj=you_object_pipeline, transform=t)
t.translate(...)
t.rotate(...)
```

Sorry for the delay, to address some of your questions:

* Transforms are applied in 2 different ways: ray traced objects (including NeRFs), actually apply the inverse transform on the rays. For rasterized objects, the model matrix from the transform object is applied in conjunction with the camera view matrix (for example, for data layers, that happens when the shaders are invoked, i.e. the PrimitivesPainter for data layers).

* Transforms are applied in a separate way to training. Training should always take place in normalized object coordinates (-1 to 1, for example). Unless you have a good reason, my suggestion would be to stick with those bounds and transform your object later for visualization purposes.

* Loading several configs: we've revamped the mains / config parser to make them easier to customize with your logic. This should be pushed soon, I plan to follow up with an example showing how to load objects via a main and view them side by side. 

* Occlusions of different neural fields / pipelines: We handle it right now by predicting a depth channel, so assuming your SDF and NeRF both produce RenderBuffers with depth channels (should do that by default, according to the behavior of our default tracers), you should be fine. Some quality issues like "z fighting" or properly blending transparent areas from the 2 fields could be tricky to get right. The depth precision you had in wisp could, i.e. account for the differences to the scene quality via other software (tuning hyperparameters and using regluarizers could mitigate that).

* Merging together various object types: I'd argue this is still an open research problem. The solution we have for multiple pipelines is sort of a "frankenrender", it works but of course not optimal.

**Comment by 3a1b2c3** [comment]
Fantastic update

**Comment by katSchmid** [comment]
ah, i realise i never posted this here: so the alpha channel looks systematically wrong in this in ways other nerf renders dont have. I only tried the interactive renderer 

[image](https://user-images.githubusercontent.com/65633379/199375227-18973ca2-64d4-472d-8381-68238b24bd53.png)
![image](https://user-images.githubusercontent.com/65633379/199375269-49a426ff-e3cf-4ad2-8490-40cd9a2d346f.png)

**Comment by 3a1b2c3** [comment]
![20221020_074148.jpg](https://user-images.githubusercontent.com/74843139/199375741-01de2d93-3790-467e-8dde-dd75efc73679.jpg)

The depth channel in my test incorrect in ways other nerf renderer seem to not show, see nerf studio above. I only tried the interactive renderer. Would you expect that to be the result?

**Comment by 3a1b2c3** [comment]
![Screenshot_20221018-222519_Docs.jpg](https://user-images.githubusercontent.com/74843139/199376011-4297860c-2366-4e26-ad62-88cd7cb31b6c.jpg) the depth channel in my test looks systemetically incorrect in ways other nerf render seems to not work (see above) Would you consider this an expected result?

**Comment by 3a1b2c3** [comment]
instant ngp depth channel 
[dsd1140310](https://insightuas.prpellr.com/p/viewer/site/prf2dd72b7/dataset/dsd1140310/surveys)

**Comment by katSchmid** [comment]
thanks, i uess good to close

**Comment by 3a1b2c3** [comment]
I think good close, thanks everybody for the help

**Comment by ttsesm** [comment]
@3a1b2c3 (or @orperel ) would be easy to elaborate a bit how it is possible to have multiple pipelines rendered at the same time. For example I would be interested to have multiple sdfs rendered/trained at the same time, is that possible?

Thanks.

**Comment by 3a1b2c3** [comment]
Oh that was a while ago. Maybe there are better ways now

**Comment by ttsesm** [comment]
> Oh that was a while ago. Maybe there are better ways now
>
Maybe @orperel can give some feedback on this then.

**Comment by ttsesm** [comment]
> If you have further questions please open a separate issue, it makes it easier for me to track :)
>
Ok, I will open a new issue
